### PR TITLE
feat(servicecatalog): Add Product Stack Asset Support

### LINKED ISSUE #20690: (servicecatalog): ProductStack does not support Assets
### Describe the feature

The feature is an improvement to the existing `ProductStack` construct to add support for the use of asset files.

### Use Case

I'm always frustrated as a Service Catalog administrator when I try to add a Lambda function to my ProductStack in CDK because I want to reference my Lambda code from an asset file, and CDK throws an error when I attempt to synthesize this. This limitation means that I'm unable to make use of ProductStack when I want to create a Service Catalog product consisting of Lambas that run large amounts of code. This is an example of a product which I would like to deploy to Service Catalog and share with end users across AWS accounts:

```
class ServerlessProduct extends sc.ProductStack {
  constructor(scope: cdk.Construct, id: string) {
    super(scope, id);

    // Defines an AWS Lambda resource
    const myHandler = new lambda.Function(this, 'Handler', {
      runtime: lambda.Runtime.NODEJS_14_X,
      code: lambda.Code.fromAsset(path.join(__dirname, 'handler')),
      handler: 'index.handler'
    });

    // Defines an API Gateway REST API resource backed by the handler function
    new apigw.LambdaRestApi(this, 'Endpoint', {
      handler: myHandler
    });
  }
}
```
`cdk synth`

Error thrown: `Service Catalog Product Stacks cannot use Assets`

### Proposed Solution

Design as of 6/9/22 by @wanjacki, @mackalex

Currently, CDK vends an asset bucket during bootstrap-time to the customer's AWS account. This bucket can be used successfully for enabling file asset support in ProductStack with CFN outputs from the parent stack for both the S3 bucket name and object key. The major issue with this approach is that when sharing a Service Catalog portfolio across accounts, a product that makes use of file assets cannot be provisioned since the parent stack with the aforementioned outputs does not exist in the end-user account.

To solve this, we could implement the usage of a bespoke S3 Bucket to contain asset files from assets used in a Service Catalog ProductStack. The bucket could exist at the Service Catalog Portfolio level which is instantiated in a ProductStack's parent stack. A bespoke bucket for this use case allows us to control the naming of the bucket as well as its permissions. Controlling the bucket name is important, especially at synth-time, since this will be referenced by resources that use assets, such as a Lambda function which references Python code stored in an asset file in the S3 bucket. Controlling permissions on a bucket which contains assets is important within the framework of Service Catalog since the administrator of a Service Catalog portfolio shares this portfolio across AWS accounts with end users who make use of products which reference asset files.

### Other Information

Additional design considerations:
We have considered making use of the bootstrap bucket which CDK vends to customers to hold assets used by resources in a ProductStack. This presents issues with cross-account sharing of Service Catalog portfolios since the assets bucket would require permissions for the end-user account to access an asset file used by a Provisioned Product, and appending permissions to an existing S3 bucket policy is nearly infeasible without overwriting the bucket policy. This is not a desirable experience, especially when the bucket policy being overwritten would be the bootstrap bucket used by many components of CDK.

Relates issues:
https://github.com/aws/aws-cdk/issues/20361

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.24.1 (build 585f9ca)

### Environment details (OS name and version, etc.)

macOS Big Sur Version 11.6.5 (20G527)

### PR DESCRIPTION
**Pull request by wanjacki**
Currently Assets are not supported in Product Stacks. Service Catalog has an unique use case where assets need to be shared cross account and sharing the entire CDK asset bucket is not ideal. Users can either create their own ProductStackAssetBucket or have one automatically generated for them based on their account Id and region. By using S3 Deployments we able to copy the assets to that bucket and share it when a portfolio is shared in Service Catalog.

More details can be found here: #20690. Closes #20690 
 
----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

----
Co-authored-by: Theron Mansilla[[imanolympic](https://github.com/imanolympic)]

### CODE REVIEW

**Comment by padaszewski** [comment]
Hi,
Thanks for the great job.
I was comparing the new API with our implementation and I didn't found any way to incorporate the own ```ProductStackAssetBucket``` into the ```ProductStackHistory``` construct.
Do You plan to provide this option?

**Comment by wanjacki** [comment]
@padaszewski That should automatically be supported.
ie. if you want to specify your own bucket
```
    const productStackHistory = new ProductStackHistory(this, 'ProductStackHistory', {
      productStack: new HelloServerlessProduct(this, 'HelloServerlessProduct', {
        assetBucket: myBucket,}),
      currentVersionName: 'v1',
      currentVersionLocked: true
      }
    )
```

**Comment by padaszewski** [comment]
@wanjacki 
Ah right, this is a property of the ProductStack. After a quick read of the README I was convinced that the bucket is passed like this ```servicecatalog.CloudFormationTemplate.fromProductStack(productStack, bucket)```. My bad, sorry.

File: packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts [hunk 1]
```diff
@@ -1,7 +1,8 @@
-import * as iam from '@aws-cdk/aws-iam';
-import * as sns from '@aws-cdk/aws-sns';
-import * as cdk from '@aws-cdk/core';
-import { Construct } from 'constructs';
+import { Construct, IConstruct } from 'constructs';
+import * as iam from '../../aws-iam';
+import { IBucket } from '../../aws-s3';
+import * as sns from '../../aws-sns';
+import * as cdk from '../../core';
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts:5]: why these changes to the import statements? This is not the normal CDK pattern.

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  public _getBucket(): IBucket {
+    return this.bucket;
+  }
+
+  /**
+   * Generate unique name for S3 bucket.
+   *
+   * @internal
+   */
+  private generateBucketName(id: string): string {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:69]: private methods do not need to be marked `@internal`. `@internal` is for APIs that must be `public` for internal usage (from another construct, for example) but we do not want actually exposed for customer usage.

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    return { bucketName, objectKey, httpUrl, s3ObjectUrl, s3Url: httpUrl };
+  }
+
+  /**
+   * Deploy all assets to S3.
+   *
+   * @internal
+   */
+  private _deployAssets() {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:104]: Same here. Do a once-over and make sure that all private methods don't have `@internal`. We also don't need to add `_` to the name of these private methods.

File: packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-synthesizer.ts [hunk 4]
```diff
@@ -16,7 +25,17 @@ export class ProductStackSynthesizer extends cdk.StackSynthesizer {
   }
 
   public addFileAsset(_asset: cdk.FileAssetSource): cdk.FileAssetLocation {
-    throw new Error('Service Catalog Product Stacks cannot use Assets');
+    if (!this.stack) {
+      throw new Error('You must call bindStack() first');
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-synthesizer.ts:29]: this message will reach the user. Do you mean to tell your user to call `bindStack()`?

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import { Construct, IConstruct } from 'constructs';
+import { hashValues } from './private/util';
+/**
+ * Product stack asset bucket props.
+ */
+export interface ProductStackAssetBucketProps {
+  /**
+   * Name of s3 asset bucket deployed
+   *
+   * @default - None
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:13]: The default is not none... a bucket name is generated if this property is not supplied.

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack.ts [hunk 6]
```diff
@@ -1,11 +1,25 @@
...
 import { ProductStackHistory } from './product-stack-history';
 
+/**
+ * Product stack props.
+ */
+export interface ProductStackProps {
+  /**
+   * Product stack asset bucket.
+   *
+   * @default - None
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product-stack.ts:18]: What does default - none mean? No asset bucket created?

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack.ts [hunk 7]
```diff
@@ -1,11 +1,25 @@
...
 import * as cdk from '@aws-cdk/core';
 import { Construct } from 'constructs';
 import { ProductStackSynthesizer } from './private/product-stack-synthesizer';
+import { ProductStackAssetBucket } from './product-stack-asset-bucket';
 import { ProductStackHistory } from './product-stack-history';
 
+/**
+ * Product stack props.
+ */
+export interface ProductStackProps {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product-stack.ts:14]:
```suggestion
export interface ProductStackOptions {
```

Options since they are all optional

File: packages/@aws-cdk/aws-servicecatalog/lib/product.ts [hunk 8]
```diff
@@ -172,6 +180,13 @@ export class CloudFormationProduct extends Product {
   public readonly productArn: string;
   public readonly productId: string;
 
+  /**
+   * The asset bucket of a product created via product stack.
+   * @param
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product.ts:185]:
```suggestion
```

File: packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts [hunk 9]
```diff
@@ -236,6 +243,20 @@ abstract class PortfolioBase extends cdk.Resource implements IPortfolio {
     }
   }
 
+  /**
+   * Gives access to Asset Buckets to Shared Accounts.
+   *
+   * @internal
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts:249]: unnecessary `@internal` tag

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    } else {
+      this.bucketName = this.generateBucketName(id);
+    }
+
+    this.bucket = new Bucket(scope, `${id}S3Bucket`, {
+      bucketName: this.bucketName,
+      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
+      encryption: BucketEncryption.KMS,
+      removalPolicy: cdk.RemovalPolicy.DESTROY,
+      autoDeleteObjects: true,
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:41]: are these properties all necessary to be hardcoded or should we expose some of them through `ProductStackAssetBucketProps`

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+      throw new Error('CDK Account ID must be defined in the application environment');
+    }
+    if (cdk.Token.isUnresolved(region)) {
+      throw new Error('CDK Region must be defined in the application environment');
+    }
+    return `product-stack-asset-bucket-${accountId}-${region}-${hashValues(id)}`;
+  }
+
+  /**
+   * Fetch the expected S3 location of an asset.
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:82]: this is doing more than fetch right? adds an asset to `this.assets`

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+      bucketName: this.bucketName,
+      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
+      encryption: BucketEncryption.KMS,
+      removalPolicy: cdk.RemovalPolicy.DESTROY,
+      autoDeleteObjects: true,
+    });
+
+    this.assets = [];
+
+    cdk.Aspects.of(this).add({
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:46]: what is the reasoning behind using `cdk.Aspects` to run deploy assets? not questioning, just wondering.

**Review by kaizencc — DISMISSED** [review]
Hi @wanjacki! Given this a quick read through with some initial comments.

File: packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-synthesizer.ts (hunk 4 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-synthesizer.ts:29]: This is the error we throw in our `synthesize` method as well. I'm not sure about this, I don't think this is an error the user can act on regardless. I don't think we can ever reach this state, so maybe this error message is just for our own sanity.

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack.ts (hunk 6 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/product-stack.ts]: Yes I updated it to be slightly more descriptive, but the asset bucket is only generated if it is needed, otherwise there is no asset bucket.

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack.ts (hunk 7 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/product-stack.ts:14]: I was unable to make this change. I believe it has to be called Props regardless.
I get the following errors on build otherwise:
```
error: [awslint:props-struct-name:@aws-cdk/aws-servicecatalog.IProductStack] all constructs must have a props struct 
error: [awslint:construct-ctor-props-type:@aws-cdk/aws-servicecatalog.ProductStack] construct "props" type should use the props struct @aws-cdk/aws-servicecatalog.ProductStackProps

```

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts (hunk 10 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:41]: I think we need these configurations to make asset buckets work as we intend for our use case.  I think allowing customers to pass their own props might add further complication and could break the intended use case. There might be some customization that customers might want done on the asset bucket, but I don't think they are needed right now.

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts (hunk 11 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:80]: Updated to be a bit more descriptive.

File: packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts (hunk 12 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts:46]: It can be done without Aspects I believe, but Aspects fits the use case and allows us to cleanly implement our use case as well as abstract the additional code to the new `product-stack-asset-bucket` constuct rather than modifying and adding additional code to  `product-stack`.

File: packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts [hunk 13]
```diff
@@ -236,6 +243,19 @@ abstract class PortfolioBase extends cdk.Resource implements IPortfolio {
...
   }
 
+  /**
+   * Gives access to Asset Buckets to Shared Accounts.
+   *
+   */
+  protected addBucketPermissionsToSharedAccounts() {
+    if (this.sharedAccounts.length > 0) {
+      for (const bucket of this.assetBuckets) {
+        bucket.grantRead(new iam.CompositePrincipal(...this.sharedAccounts.map(account => new iam.AccountPrincipal(account))),
```
Comment (corymhall) [inline packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts:253]: How are you granting access on the other side? The IAM role that gets from these buckets needs
access to be granted access to the bucket in its princial policy.

File: packages/@aws-cdk/aws-servicecatalog/test/integ.product.ts [hunk 14]
```diff
@@ -1,8 +1,9 @@
 import * as path from 'path';
```
Comment (corymhall) [inline packages/@aws-cdk/aws-servicecatalog/test/integ.product.ts:1]: This integration test only tests half of the scenario. It confirms that we can publish to the
bucket, but how do we know that we can then deploy one of these products?

**Review by corymhall — DISMISSED** [review]
I have a couple of big concerns with this PR.

1. We need an integration test that tests creating a product with assets _and_ provisioning a
   product with assets.
2. I'm not sure about how you are currently handling the asset buckets. Each individual product will
   get it's own asset bucket? What about having a single bucket that the user needs to create
   themselves? A bucket per product seems like it will lead to more work on the consumer side (if
   you want to provision product ABC then add a policy to your role with access to bucket ABC).

**Comment by wanjacki** [comment]
> I have a couple of big concerns with this PR.
> 
>     1. We need an integration test that tests creating a product with assets _and_ provisioning a
>        product with assets.
> 
>     2. I'm not sure about how you are currently handling the asset buckets. Each individual product will
>        get it's own asset bucket? What about having a single bucket that the user needs to create
>        themselves? A bucket per product seems like it will lead to more work on the consumer side (if
>        you want to provision product ABC then add a policy to your role with access to bucket ABC).

1) We can't provision a product directly in CDK and there has been no precedent for having to provision a product in any of the other integration test.
2) Yes we are limited to one asset bucket Product Stack if we are generating for the user. It is not trivial to implement it on a portfolio level and the asset itself are tied to Products not Portfolios. As a workaround we allow users to create their own asset bucket, this asset bucket can be used for multiple Product Stacks and bypasses this restriction. We grant users that the admin shares their portfolio/product with permissions to read the Bucket. Any further policy that might be needed would have to be setup and managed by the Admin (Hub) to Shared Accounts (Spoke) anyways.

File: packages/@aws-cdk/aws-servicecatalog/lib/cloudformation-template.ts [hunk 15]
```diff
@@ -46,9 +47,16 @@ export abstract class CloudFormationTemplate {
...
   /**
-    * The http url of the template in S3.
-    */
+   * The http url of the template in S3.
+   */
   readonly httpUrl: string;
+
+  /**
+   * The S3 bucket containing product stack assets.
+   * @default - None
```
[the remark this answers is not in the document]
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/cloudformation-template.ts:56]: This class is the result of binding a Artifact Template into a Product.
In the event that it is not binding to a Product Stack (URL or local asset template instead), there is no asset bucket and no asset support.
In the event that they are binding a Product Stack, if they are not using assets, then the asset bucket will not be generated and will be undefined and there will be no asset bucket.
In any other case there is an asset bucket.

File: packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-synthesizer.ts [hunk 16]
```diff
@@ -16,7 +25,17 @@ export class ProductStackSynthesizer extends cdk.StackSynthesizer {
...
+    }
+
+    if (!this.assetBucket) {
+      const parentStack = (this.stack as ProductStack)._getParentStack();
+      this.assetBucket = new ProductStackAssetBucket(parentStack, `ProductStackAssetBucket${hashValues(this.stack.stackName)}`);
+    }
```
[the remark this answers is not in the document]
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-synthesizer.ts:35]: Do you mean we create a ProductStackAssetBucket regardless of if the user is using an asset or not.
I don't that is a good idea since we would be creating resources that will not be needed (S3-deployment also deploys a lambda). Furthermore, a user could be unintentionally create a bunch of S3 Buckets (since it is one per product-stack) and fill up their S3 Bucket limit capacity.

File: packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts (hunk 13 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts:253]: Addressed in my other comment, but there is not much more we can do. We made it as easy as possible but we don't have access to the accounts it is being shared with, it is up to the Admin to configure these spoke accounts.

File: packages/@aws-cdk/aws-servicecatalog/test/integ.product.ts (hunk 14 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/test/integ.product.ts:1]: The same can be said about any of the other existing integration test. There exist validation in Service Catalog to validate the template before it can be created but it can't check everything. We can't provision the product in CDK code and even then trying to provision it would probably fall in the scope of a provision-product construct.
I'm not sure if is the products job to check if a Product is provisionable, Service Catalog allows us to create a Product regardless of if its deployable or not. We also do our best in unit test to check the generated template is as expected.

Comment (corymhall) [reply packages/@aws-cdk/aws-servicecatalog/test/integ.product.ts:1]: This is different though than existing functionality. In this PR we are adding functionality outside of service catalog. This PR is completely useless if the consumer cannot access the assets in the bucket and we have no test to assert that.

**Comment by corymhall** [comment]
@wanjacki I would recommend going through the [RFC process](https://github.com/aws/aws-cdk-rfcs/) for this feature. There are a lot of design considerations that we are not yet agreed on. The biggest thing for me is that we _have_ to consider the entire experience (publishing & consuming).

**Review by corymhall — CHANGES_REQUESTED** [review]
Putting back in request changes.

File: packages/@aws-cdk/aws-servicecatalog/test/integ.product.ts (hunk 14 above)
Comment (wanjacki) [reply packages/@aws-cdk/aws-servicecatalog/test/integ.product.ts:1]: Hey Cory, I refactored ProductStackAssetBucket to extend S3.Bucket as suggested as well as implemented the other comments.
I also added additional unit test to check Bucket Name and Bucket Policy Permissions are correct in template when shared.
If we can catch that the bucket is correct and the permissions are correct then the product should be deployable and the bucket should be acessable. If this is still not enough, please advise what what other test we can add (I don't think provisioning it is an option).

**Comment by TheRealAmazonKendra** [comment]
> @wanjacki I would recommend going through the [RFC process](https://github.com/aws/aws-cdk-rfcs/) for this feature. There are a lot of design considerations that we are not yet agreed on. The biggest thing for me is that we _have_ to consider the entire experience (publishing & consuming).

+1 on this comment. I'm going to convert this into a draft until we have an approved RFC for this.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Putting in changes requested to accurately reflect status.
